---
title: Single-cell Workshop 2021 - 03 - ç»†èƒç›¸äº’ä½œç”¨åˆ†æ
author: Shixiang Wang
date: 2023-07-31
categories: [note,scRNA-seq,R]
draft: false
description: ''

toc: true

format:
  html:
    code-fold: false
    code-overflow: wrap
    code-tools: false
---

å­¦ä¹ èµ„æ–™ï¼š[GitHub åœ°å€](https://github.com/pkuerten/single_cell_interaction.github.io/blob/main/index.md)

## å°´å°¬äº†ï¼Œæµç¨‹è·‘ä¸é€š

è¿™ä¸ªæ•™ç¨‹åœ¨è¿è¡Œåˆ° `predict_ligand_activities()` å°±å‡ºç°äº†é—®é¢˜ã€‚åé¢å°±èµ°ä¸ä¸‹å»äº†ã€‚

```
predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_
    ligands = potential_ligands)
Error in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position) : 
  all genes have same response
```

æˆ‘è‡ªå·±ä¹Ÿæ‡’å¾—æŠ˜è…¾äº†ï¼Œæœ¬èº«å°±æ˜¯æƒ³ç®€å•å­¦ä¹ äº†è§£æµç¨‹ï¼Œèµ°ä¸é€šå°±ä¸èµ°äº†ã€‚å¦‚æœæœ‰è¯»è€…çœ‹åˆ°è¿™ä¸ªæ–‡ç« ï¼Œæ„Ÿå…´è¶£å¯ä»¥é’»ç ”ä¸‹ã€‚

æˆ‘æŸ¥äº†ä¸‹ï¼Œ[Comparison of methods and resources for cell-cell communication inference from single-cell RNA-Seq data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9184522/) æ˜¾ç¤ºCellChatã€CellPhoneDBå’ŒSingleCellSignalRï¼Œå¯¹æ•°æ®å’Œèµ„æºä¸­çš„å™ªå£°éƒ½ç›¸å½“ç¨³å¥ã€‚å› è€ŒNicheNetå¹¶ä¸åœ¨æ¨èèŒƒå›´å†…ã€‚

ä¸è¿‡ä¹Ÿæåˆ°ï¼š

> Methods that additionally infer intracellular processes, such as NicheNet19, Cytotalk22, and SoptSC20 are not directly comparable but instead provide complementary analyses.

è¯´æ˜ç›®å‰çœ‹ï¼Œè¿˜æ˜¯æœ‰ä¸€äº›ç‰¹å®šçš„åˆ†ææœ‰å¯å–ä¹‹å¤„ã€‚

## æ•°æ®ä¸‹è½½

æˆ‘ä»¬å°†ä½¿ç”¨SeuratDataè½¯ä»¶åŒ…æä¾›çš„PBMCæ•°æ®é›†ã€‚è¯¥æ•°æ®é›†åŒ…å«ä¸¤ä¸ªPBMCæ ·æœ¬ï¼šç»è¿‡IFN Betaå¤„ç†çš„åˆºæ¿€æ ·æœ¬å’Œå¯¹ç…§æ ·æœ¬ã€‚

å¯¹äºç»†èƒäº¤äº’åˆ†æï¼Œæˆ‘ä»¬å°†ä½¿ç”¨NicheNetï¼Œå®ƒéœ€è¦ä¸€äº›ç²¾å¿ƒç­–åˆ’çš„æ•°æ®é›†ã€‚è™½ç„¶è¿™äº›æ•°æ®é›†å¯ä»¥åœ¨Rç¯å¢ƒå†…ä¸‹è½½ï¼Œä½†å¯¹æˆ‘æ¥è¯´ä¸‹è½½é€Ÿåº¦å¤ªæ…¢äº†ã€‚è¿™äº›æ–‡ä»¶å¯ä»¥ä»Zenodoä¸‹ï¼ˆ<https://zenodo.org/record/3260758>ï¼‰è½½ï¼šgr_network.rdsã€ligand_target_matrix.rdsã€lr_network.rdså’Œweighted_networks.rdsã€‚

è¯·å°†è¿™äº›æ•°æ®é›†æ”¾åˆ°ä½ çš„å·¥ä½œç›®å½•ä¸­ã€‚

## å‡†å¤‡

æœ¬æ–‡å‚è€ƒ Seurat [æ•°æ®æ•´åˆæ–‡æ¡£è¿›è¡Œåˆ†æçš„è¾“å…¥æ•°æ®å‡†å¤‡](https://satijalab.org/seurat/articles/integration_introduction.html)ï¼Œå‡è®¾ä½ å¯¹æ•°æ®é¢„å¤„ç†éƒ¨åˆ†å·²ç»å¾ˆç†Ÿæ‚‰ã€‚

## å®‰è£…åŒ…

å®‰è£…æ‰€éœ€è¦çš„åŒ…ã€‚

```R
BiocManager::install("limma")
BiocManager::install("glmGamPoi")
# install.packages("devtools")
install.packages("https://seurat.nygenome.org/src/contrib/ifnb.SeuratData_3.0.0.tar.gz", repos = NULL, type = "source") 
devtools::install_github("saeyslab/nichenetr")
# å¯ä»¥ä¸‹è½½ä¹‹åå®‰è£…ï¼Œé¿å…å®‰è£…ä¸ä¸Š
# https://github.com/saeyslab/nichenetr
# unzip("~/Downloads/nichenetr-master.zip")
# ç¼ºä»€ä¹ˆè£…ä»€ä¹ˆ
# BiocManager::install(c("fdrtool", "DiagrammeR", "mlrMBO", "parallelMap", "emoa", "DiceKriging"))
# install.packages("nichenetr-master", repos = NULL, type = "source")
install.packages("tidyverse")
```

åŠ è½½åŒ…ï¼š

```{r}
# load into your session
library(ifnb.SeuratData)
library(Seurat)

# load dataset
data("ifnb")
```

## è¯»å–æ•°æ®é›†å¹¶åˆ›å»ºä¸€ä¸ªSeuratå¯¹è±¡

ç¬¬ä¸€æ­¥æ˜¯ä»SeuratDataè½¯ä»¶åŒ…ä¸­è¯»å–æ•°æ®ã€‚è¾“å…¥æ•°æ®æ ¹æ®æ ·æœ¬è¿›è¡Œæ‹†åˆ†ï¼Œå¹¶åˆ†åˆ«åº”ç”¨SCtranformã€‚è¯·æ³¨æ„ï¼Œè¿™å–ä»£äº†NormalizeDataã€ScaleDataå’ŒFindVariableFeaturesçš„æ­¥éª¤ã€‚SCTransformæ˜¯ä¸€ç§ä¸“é—¨è®¾è®¡ç”¨äºå•ç»†èƒUMIè®¡æ•°æ•°æ®çš„ç»Ÿè®¡æ–¹æ³•ã€‚å®ƒå…‹æœäº†ä¹‹å‰æ‰¹é‡è®¾è®¡çš„å½’ä¸€åŒ–æ–¹æ³•ä¸­ä¸€äº›è¿‡åº¦æ‹Ÿåˆçš„é™åˆ¶ã€‚æœ‰å…³å…¶ä¼˜åŠ¿çš„æ›´å¤šè¯¦ç»†ä¿¡æ¯å¯ä»¥åœ¨[SCtranformçš„è®ºæ–‡](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1)ä¸­æ‰¾åˆ°ã€‚è¯·æ³¨æ„ï¼Œæˆ‘ä»¬æ²¡æœ‰åº”ç”¨ä»»ä½•è´¨æ§æ­¥éª¤ã€‚è¿™æ˜¯å› ä¸ºæ•°æ®å·²ç»é¢„å…ˆæ¸…æ´ï¼Œå› æ­¤æ— éœ€é‡å¤è¿™äº›æ­¥éª¤ã€‚

```{r}
ifnb.list <- SplitObject(ifnb, split.by = "stim")
ifnb.list <- lapply(X = ifnb.list, FUN = SCTransform, method = "glmGamPoi")
features <- SelectIntegrationFeatures(object.list = ifnb.list, nfeatures = 3000)
ifnb.list <- PrepSCTIntegration(object.list = ifnb.list, anchor.features = features)
ifnb.list <- lapply(X = ifnb.list, FUN = RunPCA, features = features)
```

æ¥ä¸‹æ¥ï¼Œæˆ‘ä»¬å°†ä¸¤ä¸ªæ•°æ®é›†è¿›è¡Œæ•´åˆã€‚æ•´åˆä¾èµ–äºåœ¨ä¸¤ä¸ªæ•°æ®é›†ä¸­éƒ½å­˜åœ¨çš„é«˜å˜é‡ç‰¹å¾ï¼Œå¹¶ä¸”åˆ†ä¸ºä¸¤ä¸ªæ­¥éª¤ã€‚ç¬¬ä¸€æ­¥FindIntegrationAnchorså®šä¹‰é”šç‚¹æˆ–ç»†èƒå¯¹ï¼ˆæ¯ä¸ªæ ·æœ¬ä¸­å„é€‰æ‹©ä¸€ä¸ªç»†èƒï¼‰ï¼Œè¿™äº›ç»†èƒéå¸¸ç›¸ä¼¼ï¼Œå› æ­¤å¯ä»¥æœ‰ä¿¡å¿ƒåœ°å°†å®ƒä»¬åˆ†é…åˆ°ç›¸åŒçš„ç°‡ï¼ˆç»†èƒç±»å‹å’ŒçŠ¶æ€ï¼‰ã€‚ç¬¬äºŒæ­¥IntegrateDataä½¿ç”¨æ‰€å®šä¹‰çš„é”šç‚¹å¯¹é½å®Œæ•´çš„æ•°æ®é›†ã€‚æœ‰å…³æ•´åˆæ–¹æ³•çš„æ›´å¤šè¯¦ç»†ä¿¡æ¯å’Œè‡ªåŸå§‹Seuratè®ºæ–‡ä»¥æ¥çš„æ”¹è¿›å¯ä»¥åœ¨ä»–ä»¬[æœ€è¿‘çš„å‡ºç‰ˆç‰©](https://pubmed.ncbi.nlm.nih.gov/34062119/)ä¸­æ‰¾åˆ°ã€‚

```{r}
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT", anchor.features = features, dims = 1:30, reduction = "rpca", k.anchor = 20)
immune.combined.sct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT", dims = 1:30)
```

æ•´åˆåçš„å¯¹è±¡åŒ…å«ç»è¿‡æ‰¹æ¬¡æ•ˆåº”æ ¡æ­£çš„å€¼ï¼Œä»¥åŠåŸå§‹è®¡æ•°å€¼ä½œä¸ºå•ç‹¬çš„æµ‹å®šã€‚

```{r}
immune.combined.sct <- RunPCA(immune.combined.sct, verbose = FALSE)
immune.combined.sct <- RunUMAP(immune.combined.sct, reduction = "pca", dims = 1:30)
```

å¯è§†åŒ–ï¼š

```{r}
p1 <- DimPlot(immune.combined.sct, reduction = "umap", group.by = "stim")
p2 <- DimPlot(immune.combined.sct, reduction = "umap", group.by = "seurat_annotations", label = TRUE,
    repel = TRUE)
p1 + p2
```

## ç»†èƒç›¸äº’ä½œç”¨åˆ†æ

[NicheNet](https://pubmed.ncbi.nlm.nih.gov/31819264/)æ—¨åœ¨é€šè¿‡å°†ç‰¹å®šç°‡çš„å•ç»†èƒè¡¨è¾¾æ•°æ®ä¸å…³äºé…ä½“-å—ä½“å¯¹å’Œç›®æ ‡ä¸‹æ¸¸çš„åŸºå› è°ƒæ§ç½‘ç»œçš„å…ˆéªŒçŸ¥è¯†ç›¸ç»“åˆï¼Œé¢„æµ‹é…ä½“å’Œç›®æ ‡ç»†èƒä¹‹é—´çš„ç›¸äº’ä½œç”¨è¿æ¥ã€‚ç‰¹åˆ«åœ°ï¼Œå®ƒæ—¨åœ¨å®šä¹‰èƒ½å¤Ÿæœ€å¥½è§£é‡Šç›®æ ‡ç°‡ä¸­è§‚å¯Ÿåˆ°çš„å·®å¼‚è¡¨è¾¾çš„é…ä½“ã€‚

æˆ‘ä»¬é¦–å…ˆåŠ è½½å¿…è¦çš„Rè½¯ä»¶åŒ…ã€‚âš ï¸åœ¨æˆ‘åŠ è½½äº†NicheNetè½¯ä»¶åŒ…åï¼Œéƒ¨åˆ†Seuratæ•´åˆå‘½ä»¤æœªèƒ½æˆåŠŸæ‰§è¡Œã€‚ä¸¤ä¸ªè½¯ä»¶åŒ…ä¹‹é—´å¯èƒ½å­˜åœ¨ä¸€äº›ä¸å…¼å®¹æ€§ã€‚


```{r}
library(nichenetr)
library(tidyverse)
```

åŠ è½½æ•°æ®é›†ï¼š

```{r}
ligand_target_matrix = readRDS("/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/ligand_target_matrix.rds")
ligand_target_matrix[1:5,1:5]
```


```{r}
lr_network = readRDS("/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/lr_network.rds")
head(lr_network)
```


```{r}
weighted_networks = readRDS("/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/weighted_networks.rds")
weighted_networks_lr = weighted_networks$lr_sig %>% inner_join(lr_network %>% distinct(from,to), by = c("from","to"))
head(weighted_networks$lr_sig) # interactions and their weights in the ligand-receptor + signaling network
```


```{r}
Idents(immune.combined.sct) <-immune.combined.sct$seurat_annotations
DefaultAssay(immune.combined.sct) <- "integrated"
```

åœ¨æœ¬æ–‡ä¸­ï¼Œæˆ‘ä»¬ä¾èµ–äºæ•°æ®æä¾›çš„ç­›é€‰ç»†èƒæ³¨é‡Šã€‚
æˆ‘ä»¬å°†é‡ç‚¹æ”¾åœ¨CD8 Tç»†èƒä½œä¸ºæ¥æ”¶ç»†èƒï¼Œå¹¶å¯»æ‰¾pDCã€CD14 Monoã€CD16 Monoã€NKã€Bã€DCã€B Activatedç¾¤é›†ä¸­çš„å‘é€å™¨é…ä½“ã€‚

æˆ‘ä»¬å®šä¹‰æ¥æ”¶ç»†èƒå’Œå‘é€å™¨ç»†èƒç¾¤ä½“ã€‚æ­¤å¤–ï¼Œæˆ‘ä»¬è¿˜è¯†åˆ«åœ¨æ¥æ”¶ç»†èƒä¸­è¡¨è¾¾çš„åŸºå› ï¼Œè¿™äº›åŸºå› ä¹Ÿæ˜¯æˆ‘ä»¬çŸ¥è¯†åº“çš„ä¸€éƒ¨åˆ†ã€‚

ğŸ’¡å¦‚æœæŸä¸ªåŸºå› åœ¨æ•°æ®åº“ä¸­æ²¡æœ‰è¢«çŸ¥é“ç”±ä»»ä½•é…ä½“è°ƒæ§ï¼Œè¯¥æ–¹æ³•å°†æ— æ³•åœ¨å…¶é¢„æµ‹ä¸­ä½¿ç”¨å®ƒã€‚


```{r}
receiver = "CD8 T"
expressed_genes_receiver = get_expressed_genes(receiver, immune.combined.sct, pct = 0.10, assay_oi="RNA")
background_expressed_genes = expressed_genes_receiver %>% .[. %in% rownames(ligand_target_matrix)]

sender_celltypes = c("pDC", "CD14 Mono" ,  "CD16 Mono",  "NK", "B", "DC", "B Activated")

list_expressed_genes_sender = sender_celltypes %>% unique() %>% lapply(get_expressed_genes, immune.combined.sct, 0.10, assay_oi="RNA") 
expressed_genes_sender = list_expressed_genes_sender %>% unlist() %>% unique()
```

å®šä¹‰æ¥æ”¶ç»†èƒä¸­ç›¸å¯¹äºæ ·æœ¬å¤„ç†çš„å·®å¼‚åŸºå› ã€‚


```{r}
seurat_obj_receiver= subset(immune.combined.sct, idents = receiver)
seurat_obj_receiver = SetIdent(immune.combined.sct, value = seurat_obj_receiver[["orig.ident"]])

condition_oi = "IMMUNE_STIM"
condition_reference = "IMMUNE_CTRL" 

DE_table_receiver = FindMarkers(object = seurat_obj_receiver, ident.1 = condition_oi, ident.2 = condition_reference, min.pct = 0.10, assay="RNA") %>% rownames_to_column("gene")

geneset_oi = DE_table_receiver %>% filter(p_val <= 0.05 & abs(avg_log2FC) >= 0.25) %>% pull(gene)
geneset_oi = geneset_oi %>% .[. %in% rownames(ligand_target_matrix)]

```

å®šä¹‰åœ¨å„è‡ªç¾¤é›†ä¸­è¡¨è¾¾çš„é…ä½“å’Œå—ä½“åŸºå› ã€‚æ½œåœ¨çš„é…ä½“é›†è¢«ç¼©å°ä¸ºåœ¨æ¥æ”¶ç»†èƒç¾¤é›†ä¸­è¡¨è¾¾çš„å…·æœ‰é¶æ ‡çš„é…ä½“ã€‚


```{r}
ligands = lr_network %>% pull(from) %>% unique()
receptors = lr_network %>% pull(to) %>% unique()

expressed_ligands = intersect(ligands,expressed_genes_sender)
expressed_receptors = intersect(receptors,expressed_genes_receiver)

potential_ligands = lr_network %>% filter(from %in% expressed_ligands & to %in% expressed_receptors) %>% pull(from) %>% unique()
```

NicheNetçš„é…ä½“æ´»æ€§åˆ†ææ ¹æ®æ¥æ”¶ç»†èƒä¸­å­˜åœ¨çš„è¡¨è¾¾é¶æ ‡åŸºå› çš„æƒ…å†µå¯¹é…ä½“è¿›è¡Œæ’åã€‚


```{r}
# ligand_activities = predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_ligands = potential_ligands)

# ligand_activities = ligand_activities %>% arrange(-pearson) %>% mutate(rank = rank(desc(pearson)))
# ligand_activities
```

