<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Shixiang Wang">
<meta name="dcterms.date" content="2023-07-31">
<title>Reasoning - Single-cell Workshop 2021 - 03 - 细胞相互作用分析</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<!-- Primary Meta Tags --><title>Reasoning - ShixiangWang's blog</title>
<meta name="title" content="Reasoning - ShixiangWang's blog">
<meta name="description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<!-- Open Graph / Facebook --><meta property="og:type" content="website">
<meta property="og:url" content="https://shixiangwang.github.io/posts/welcome/">
<meta property="og:title" content="Reasoning - Single-cell Workshop 2021 - 03 - 细胞相互作用分析">
<meta property="og:description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<meta property="og:image" content="https://avatars.githubusercontent.com/u/25057508?v=4">
<!-- Twitter --><meta property="twitter:card" content="summary_large_image">
<meta property="twitter:url" content="https://shixiangwang.github.io/posts/welcome/">
<meta property="twitter:title" content="Reasoning - ShixiangWang's blog">
<meta property="twitter:description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<meta property="twitter:image" content="https://avatars.githubusercontent.com/u/25057508?v=4">
<meta name="referrer" content="no-referrer">
<link rel="stylesheet" href="../../styles.css">
<meta name="twitter:title" content="Reasoning - Single-cell Workshop 2021 - 03 - 细胞相互作用分析">
<meta name="twitter:description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<meta name="twitter:creator" content="@WangShxiang">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Reasoning</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">
<li>
    <a class="dropdown-item" href="../../resource.html">
 <span class="dropdown-text">Code and Books</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://shixiangwang.github.io/cv-shixiang/"> 
<span class="menu-text">Resume</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://shixiangwang.github.io/home/logo/qrcode.jpg"> <i class="bi bi-wechat" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.zhihu.com/people/shixiangwang"> <i class="bi bi-quora" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ShixiangWang"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/WangShxiang"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Single-cell Workshop 2021 - 03 - 细胞相互作用分析</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">note</div>
                <div class="quarto-category">scRNA-seq</div>
                <div class="quarto-category">R</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Shixiang Wang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">July 31, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#%E5%B0%B4%E5%B0%AC%E4%BA%86%E6%B5%81%E7%A8%8B%E8%B7%91%E4%B8%8D%E9%80%9A" id="toc-尴尬了流程跑不通" class="nav-link active" data-scroll-target="#%E5%B0%B4%E5%B0%AC%E4%BA%86%E6%B5%81%E7%A8%8B%E8%B7%91%E4%B8%8D%E9%80%9A">尴尬了，流程跑不通</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD" id="toc-数据下载" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD">数据下载</a></li>
  <li><a href="#%E5%87%86%E5%A4%87" id="toc-准备" class="nav-link" data-scroll-target="#%E5%87%86%E5%A4%87">准备</a></li>
  <li><a href="#%E5%AE%89%E8%A3%85%E5%8C%85" id="toc-安装包" class="nav-link" data-scroll-target="#%E5%AE%89%E8%A3%85%E5%8C%85">安装包</a></li>
  <li><a href="#%E8%AF%BB%E5%8F%96%E6%95%B0%E6%8D%AE%E9%9B%86%E5%B9%B6%E5%88%9B%E5%BB%BA%E4%B8%80%E4%B8%AAseurat%E5%AF%B9%E8%B1%A1" id="toc-读取数据集并创建一个seurat对象" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96%E6%95%B0%E6%8D%AE%E9%9B%86%E5%B9%B6%E5%88%9B%E5%BB%BA%E4%B8%80%E4%B8%AAseurat%E5%AF%B9%E8%B1%A1">读取数据集并创建一个Seurat对象</a></li>
  <li><a href="#%E7%BB%86%E8%83%9E%E7%9B%B8%E4%BA%92%E4%BD%9C%E7%94%A8%E5%88%86%E6%9E%90" id="toc-细胞相互作用分析" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E7%9B%B8%E4%BA%92%E4%BD%9C%E7%94%A8%E5%88%86%E6%9E%90">细胞相互作用分析</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><p>学习资料：<a href="https://github.com/pkuerten/single_cell_interaction.github.io/blob/main/index.md">GitHub 地址</a></p>
<section id="尴尬了流程跑不通" class="level2"><h2 class="anchored" data-anchor-id="尴尬了流程跑不通">尴尬了，流程跑不通</h2>
<p>这个教程在运行到 <code>predict_ligand_activities()</code> 就出现了问题。后面就走不下去了。</p>
<pre><code>predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_
    ligands = potential_ligands)
Error in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position) : 
  all genes have same response</code></pre>
<p>我自己也懒得折腾了，本身就是想简单学习了解流程，走不通就不走了。如果有读者看到这个文章，感兴趣可以钻研下。</p>
<p>我查了下，<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9184522/">Comparison of methods and resources for cell-cell communication inference from single-cell RNA-Seq data</a> 显示CellChat、CellPhoneDB和SingleCellSignalR，对数据和资源中的噪声都相当稳健。因而NicheNet并不在推荐范围内。</p>
<p>不过也提到：</p>
<blockquote class="blockquote">
<p>Methods that additionally infer intracellular processes, such as NicheNet19, Cytotalk22, and SoptSC20 are not directly comparable but instead provide complementary analyses.</p>
</blockquote>
<p>说明目前看，还是有一些特定的分析有可取之处。</p>
</section><section id="数据下载" class="level2"><h2 class="anchored" data-anchor-id="数据下载">数据下载</h2>
<p>我们将使用SeuratData软件包提供的PBMC数据集。该数据集包含两个PBMC样本：经过IFN Beta处理的刺激样本和对照样本。</p>
<p>对于细胞交互分析，我们将使用NicheNet，它需要一些精心策划的数据集。虽然这些数据集可以在R环境内下载，但对我来说下载速度太慢了。这些文件可以从Zenodo下（<a href="https://zenodo.org/record/3260758" class="uri">https://zenodo.org/record/3260758</a>）载：gr_network.rds、ligand_target_matrix.rds、lr_network.rds和weighted_networks.rds。</p>
<p>请将这些数据集放到你的工作目录中。</p>
</section><section id="准备" class="level2"><h2 class="anchored" data-anchor-id="准备">准备</h2>
<p>本文参考 Seurat <a href="https://satijalab.org/seurat/articles/integration_introduction.html">数据整合文档进行分析的输入数据准备</a>，假设你对数据预处理部分已经很熟悉。</p>
</section><section id="安装包" class="level2"><h2 class="anchored" data-anchor-id="安装包">安装包</h2>
<p>安装所需要的包。</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"https://seurat.nygenome.org/src/contrib/ifnb.SeuratData_3.0.0.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span> </span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"saeyslab/nichenetr"</span><span class="op">)</span></span>
<span><span class="co"># 可以下载之后安装，避免安装不上</span></span>
<span><span class="co"># https://github.com/saeyslab/nichenetr</span></span>
<span><span class="co"># unzip("~/Downloads/nichenetr-master.zip")</span></span>
<span><span class="co"># 缺什么装什么</span></span>
<span><span class="co"># BiocManager::install(c("fdrtool", "DiagrammeR", "mlrMBO", "parallelMap", "emoa", "DiceKriging"))</span></span>
<span><span class="co"># install.packages("nichenetr-master", repos = NULL, type = "source")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>加载包：</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load into your session</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ifnb.SeuratData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Attaching SeuratObject</code></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="读取数据集并创建一个seurat对象" class="level2"><h2 class="anchored" data-anchor-id="读取数据集并创建一个seurat对象">读取数据集并创建一个Seurat对象</h2>
<p>第一步是从SeuratData软件包中读取数据。输入数据根据样本进行拆分，并分别应用SCtranform。请注意，这取代了NormalizeData、ScaleData和FindVariableFeatures的步骤。SCTransform是一种专门设计用于单细胞UMI计数数据的统计方法。它克服了之前批量设计的归一化方法中一些过度拟合的限制。有关其优势的更多详细信息可以在<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">SCtranform的论文</a>中找到。请注意，我们没有应用任何质控步骤。这是因为数据已经预先清洁，因此无需重复这些步骤。</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html">SplitObject</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">ifnb.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">ifnb.list</span>, FUN <span class="op">=</span> <span class="va">SCTransform</span>, method <span class="op">=</span> <span class="st">"glmGamPoi"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cell attributes from input UMI matrix: log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Variance stabilizing transformation of count matrix of size 12747 by 6548</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Model formula is y ~ log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Get Negative Binomial regression parameters per gene</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 2000 genes, 5000 cells</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found 95 outliers - those will be ignored in fitting/regularization step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Second step: Get residuals using fitted parameters for 12747 genes</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing corrected count matrix for 12747 genes</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating gene attributes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Wall clock passed: Time difference of 31.92474 secs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Determine variable features</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Place corrected count matrix in counts slot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Set default assay to SCT</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cell attributes from input UMI matrix: log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Variance stabilizing transformation of count matrix of size 12658 by 7451</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Model formula is y ~ log_umi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Get Negative Binomial regression parameters per gene</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using 2000 genes, 5000 cells</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found 101 outliers - those will be ignored in fitting/regularization step</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Second step: Get residuals using fitted parameters for 12658 genes</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing corrected count matrix for 12658 genes</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating gene attributes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Wall clock passed: Time difference of 30.21689 secs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Determine variable features</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Place corrected count matrix in counts slot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Set default assay to SCT</code></pre>
</div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html">SelectIntegrationFeatures</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">ifnb.list</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span>
<span><span class="va">ifnb.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PrepSCTIntegration.html">PrepSCTIntegration</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">ifnb.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span>
<span><span class="va">ifnb.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">ifnb.list</span>, FUN <span class="op">=</span> <span class="va">RunPCA</span>, features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  FTL, TIMP1, IL8, CCL2, FTH1, C15orf48, S100A8, SOD2, ACTB, S100A9 
       LYZ, FCER1G, TYROBP, S100A4, HLA-DRA, CD63, CD14, S100A11, LGALS3, ANXA5 
       LGALS1, CXCL3, S100A10, IL1B, CTSL, PLAUR, CTSB, TYMP, S100A6, CST3 
Negative:  RPL3, RPS18, RPS6, LTB, RPL13, CCR7, GIMAP7, RPL13A, PTMA, CXCR4 
       RPSA, RPS2, RPL21, CD3D, RPS5, LDHB, NPM1, RPS4X, CCL5, RPS3 
       TMEM66, RPS14, RPS27, PABPC1, RPL10A, RPS15A, RPL5, RPL32, RPS19, RPL34 
PC_ 2 
Positive:  CCL2, CCL5, GNLY, S100A8, NKG7, FTL, S100A9, GZMB, GIMAP7, CD7 
       CD3D, CD14, FTH1, IL8, CXCL3, CD2, FGFBP2, CCL7, RGCC, RARRES3 
       GZMH, APOBEC3G, TMEM66, CLIC3, ITM2A, CTSL, LTB, IL32, LCK, GIMAP5 
Negative:  HLA-DRA, CD74, HLA-DQA1, HLA-DPA1, HLA-DRB1, HLA-DPB1, HLA-DQB1, CD83, TXN, CCR7 
       CCL22, HLA-DMA, IDO1, ACTB, HERPUD1, LYZ, SYNGR2, CD79A, MIR155HG, BIRC3 
       PKIB, RAB9A, SERPINB1, RAMP1, GPR183, CST3, FABP5, HLA-DRB5, TVP23A, ID2 
PC_ 3 
Positive:  RPS18, RPL13, LTB, RPS6, FTL, RPL3, RPL32, RPL10, RPS2, PABPC1 
       RPL34, RPL13A, RPS14, RPS4X, RPL21, S100A8, CCL2, RPS12, RPL11, RPL7 
       RPS3A, RPL18A, LDHB, SELL, RPS15A, RPS3, RPS8, S100A9, RPS5, RPS27A 
Negative:  GNLY, NKG7, CCL5, GZMB, CST7, APOBEC3G, FGFBP2, CLIC3, GZMH, KLRD1 
       GZMA, CTSW, PRF1, HOPX, KLRC1, CHST12, TNFRSF18, CCL4, VMO1, FCGR3A 
       DUSP2, LDHA, CXCR3, C1orf21, ID2, MATK, SPON2, SH2D1B, RARRES3, APMAP 
PC_ 4 
Positive:  HSPB1, HSP90AB1, HSPE1, CACYBP, CCL2, HSPA8, CREM, HSPD1, SRSF7, HSPH1 
       HSPA1A, FTL, CD69, UBB, GADD45B, UBC, SRSF2, ZFAND2A, HSP90AA1, IL8 
       NOP58, DDIT4, NR4A2, CHORDC1, H3F3B, BTG1, RSRC2, DNAJB1, SNHG15, YPEL5 
Negative:  VMO1, FCGR3A, TIMP1, MS4A7, RPL3, FAM26F, TNFSF10, RPS4X, RPS2, RPS19 
       RPL10, RPS18, PLAC8, CXCL16, CXCL10, HLA-DPA1, LST1, HLA-DPB1, RPL13, TNFSF13B 
       MS4A4A, RPS14, IFITM3, HN1, AIF1, GBP5, RPS6, CCL5, CD74, GBP1 
PC_ 5 
Positive:  CCL2, FTL, GNLY, CCL5, NKG7, IL8, GZMB, S100A8, S100A9, LYZ 
       HLA-DRA, CST7, CD74, HLA-DQA1, CXCL3, TXN, FGFBP2, CD63, GAPDH, CLIC3 
       IL1B, HLA-DPB1, HLA-DRB1, VIM, RPL3, GZMH, CCL7, APOBEC3G, C15orf48, HLA-DQB1 
Negative:  VMO1, FCGR3A, MS4A7, TIMP1, FAM26F, TNFSF10, CXCL10, HSPB1, IFITM3, GBP1 
       AIF1, PLAC8, MS4A4A, LST1, GBP5, CXCL16, HN1, SAT1, ATP1B3, OAS1 
       ISG15, C3AR1, WARS, CTSC, PPM1N, RGS2, TNFSF13B, SOD2, IFIT3, FGL2 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  FTL, CXCL10, CCL2, CCL8, APOBEC3A, SOD2, TIMP1, C15orf48, CCL7, APOBEC3B 
       ISG15, IDO1, CCL3, IFITM3, CXCL11, LYZ, TYROBP, CCL4, IFI27, FCER1G 
       FTH1, CTSL, SAT1, TYMP, IL1RN, LGALS3, S100A11, CTSC, CD63, LGALS1 
Negative:  RPL3, RPS6, RPS18, RPL13, RPL21, RPL7, RPL13A, CCR7, RPL10A, RPS4X 
       RPS2, GIMAP7, RPS3, RPS3A, RPS14, RPS27A, PTMA, RPL10, RPS15A, RPL34 
       RPS5, RPL32, PABPC1, RPL5, CXCR4, RPS27, RPL31, LTB, MALAT1, RPL14 
PC_ 2 
Positive:  CCL2, CCL8, CCL7, RPS18, RPL13, RPS6, FTL, RPL3, RPL32, SELL 
       RPS14, RPL34, LTB, RPL7, RPL13A, RPL21, RPL10, GIMAP7, PABPC1, RPS2 
       RPS4X, APOBEC3B, RPL10A, RPS15A, RPS3A, LDHB, S100A9, RPS12, RPL35A, RPL31 
Negative:  GNLY, GZMB, NKG7, CCL5, PRF1, CD74, CLIC3, CST7, GZMA, HLA-DRA 
       APOBEC3G, GZMH, HLA-DPB1, HLA-DPA1, KLRD1, CTSW, TXN, FCGR3A, HLA-DQA1, VMO1 
       HLA-DRB1, FASLG, ID2, FGFBP2, CD83, KLRC1, MS4A7, HLA-DQB1, RARRES3, TIMP1 
PC_ 3 
Positive:  CD74, HLA-DRA, HLA-DPB1, CXCL10, HLA-DPA1, HLA-DQA1, HLA-DRB1, TXN, TIMP1, CD83 
       VMO1, FCGR3A, CCR7, HLA-DQB1, MS4A7, MARCKSL1, IRF8, CST3, BIRC3, SYNGR2 
       MS4A4A, SERPINB1, CD79A, CXCL16, CD86, MIR155HG, HLA-DMA, CCL22, ACTB, FAM26F 
Negative:  GNLY, CCL2, CCL8, NKG7, CCL5, GZMB, CCL7, PRF1, GZMA, GZMH 
       CLIC3, KLRD1, CTSW, RARRES3, S100A9, ALOX5AP, FASLG, APOBEC3G, FGFBP2, CTSL 
       CD7, CCL4, KLRC1, CST7, S100A8, ANXA1, CCL3, CD247, IL32, LGALS3 
PC_ 4 
Positive:  FCGR3A, VMO1, MS4A7, TIMP1, MS4A4A, CXCL16, FAM26F, PLAC8, TNFSF10, TNFAIP6 
       C3AR1, LST1, FTH1, CXCL10, TYROBP, CTSC, AIF1, GBP5, PPM1N, CFD 
       MT2A, CDKN1C, SERPINA1, IFITM3, SLC31A2, HN1, PLAUR, FCER1G, FCGR3B, EDN1 
Negative:  CCL2, CCL8, CD74, HLA-DRA, TXN, HLA-DQA1, CCL7, HLA-DPB1, FTL, HLA-DRB1 
       LYZ, MIR155HG, HLA-DQB1, CD83, IRF8, FABP5, CCR7, MARCKSL1, BIRC3, APOBEC3B 
       TSPAN13, HERPUD1, HSPE1, HSPD1, HLA-DPA1, ID3, PMAIP1, HSP90AB1, CD79A, NME1 
PC_ 5 
Positive:  CD74, HLA-DPB1, HLA-DRA, RPL3, TXN, RPS18, RPS4X, MARCKSL1, HLA-DPA1, RPL10 
       RPL13, LYZ, HLA-DRB1, FTL, RPS2, GNLY, RPS14, GZMB, CCL5, RPS3 
       RPL34, NKG7, RPL11, RPL13A, CST3, RPL32, RPS6, RPL21, HLA-DQA1, LSP1 
Negative:  CCL4, CCL3, CD69, SRSF7, HSPE1, HSP90AB1, CACYBP, HSPA8, HSPD1, SRSF2 
       DDIT4, CREM, GADD45B, UBC, NOP58, UBB, HSPH1, TCP1, JUNB, SOD1 
       SAT1, CLK1, RSRC2, BTG1, DNAJB6, NR4A2, MYC, IRF1, DNAJB1, BATF </code></pre>
</div>
</div>
<p>接下来，我们将两个数据集进行整合。整合依赖于在两个数据集中都存在的高变量特征，并且分为两个步骤。第一步FindIntegrationAnchors定义锚点或细胞对（每个样本中各选择一个细胞），这些细胞非常相似，因此可以有信心地将它们分配到相同的簇（细胞类型和状态）。第二步IntegrateData使用所定义的锚点对齐完整的数据集。有关整合方法的更多详细信息和自原始Seurat论文以来的改进可以在他们<a href="https://pubmed.ncbi.nlm.nih.gov/34062119/">最近的出版物</a>中找到。</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">immune.anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html">FindIntegrationAnchors</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">ifnb.list</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span>, anchor.features <span class="op">=</span> <span class="va">features</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction <span class="op">=</span> <span class="st">"rpca"</span>, k.anchor <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing within dataset neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding all pairwise anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting new data onto SVD
Projecting new data onto SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 32565 anchors</code></pre>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">immune.combined.sct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateData.html">IntegrateData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">immune.anchors</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging dataset 1 into 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting anchors for merged samples</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Integrating data</code></pre>
</div>
</div>
<p>整合后的对象包含经过批次效应校正的值，以及原始计数值作为单独的测定。</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">immune.combined.sct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">immune.combined.sct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:39:33 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:39:33 Read 13999 rows and found 30 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:39:33 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:39:33 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
15:39:35 Writing NN index file to temp file /var/folders/bj/nw1w4g1j37ddpgb6zmh3sfh80000gn/T//RtmpF254ot/file1575b69d2520a
15:39:35 Searching Annoy index using 1 thread, search_k = 3000
15:39:38 Annoy recall = 100%
15:39:39 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
15:39:40 Initializing from normalized Laplacian + noise (using irlba)
15:39:41 Commencing optimization for 200 epochs, with 600044 positive edges
15:39:49 Optimization finished</code></pre>
</div>
</div>
<p>可视化：</p>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"seurat_annotations"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="细胞相互作用分析" class="level2"><h2 class="anchored" data-anchor-id="细胞相互作用分析">细胞相互作用分析</h2>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/31819264/">NicheNet</a>旨在通过将特定簇的单细胞表达数据与关于配体-受体对和目标下游的基因调控网络的先验知识相结合，预测配体和目标细胞之间的相互作用连接。特别地，它旨在定义能够最好解释目标簇中观察到的差异表达的配体。</p>
<p>我们首先加载必要的R软件包。⚠️在我加载了NicheNet软件包后，部分Seurat整合命令未能成功执行。两个软件包之间可能存在一些不兼容性。</p>
<div class="cell">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">nichenetr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.2     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.2     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.1     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
<p>加载数据集：</p>
<div class="cell">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ligand_target_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/ligand_target_matrix.rds"</span><span class="op">)</span></span>
<span><span class="va">ligand_target_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                CXCL1        CXCL2        CXCL3        CXCL5         PPBP
A1BG     3.534343e-04 4.041324e-04 3.729920e-04 3.080640e-04 2.628388e-04
A1BG-AS1 1.650894e-04 1.509213e-04 1.583594e-04 1.317253e-04 1.231819e-04
A1CF     5.787175e-04 4.596295e-04 3.895907e-04 3.293275e-04 3.211944e-04
A2M      6.027058e-04 5.996617e-04 5.164365e-04 4.517236e-04 4.590521e-04
A2M-AS1  8.898724e-05 8.243341e-05 7.484018e-05 4.912514e-05 5.120439e-05</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">lr_network</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/lr_network.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">lr_network</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  from  to    source         database
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;          &lt;chr&gt;   
1 CXCL1 CXCR2 kegg_cytokines kegg    
2 CXCL2 CXCR2 kegg_cytokines kegg    
3 CXCL3 CXCR2 kegg_cytokines kegg    
4 CXCL5 CXCR2 kegg_cytokines kegg    
5 PPBP  CXCR2 kegg_cytokines kegg    
6 CXCL6 CXCR2 kegg_cytokines kegg    </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">weighted_networks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"/Users/wsx/Library/CloudStorage/OneDrive-shanghaitech.edu.cn/Public/data/weighted_networks.rds"</span><span class="op">)</span></span>
<span><span class="va">weighted_networks_lr</span> <span class="op">=</span> <span class="va">weighted_networks</span><span class="op">$</span><span class="va">lr_sig</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="va">lr_network</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">from</span>,<span class="va">to</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"from"</span>,<span class="st">"to"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">weighted_networks</span><span class="op">$</span><span class="va">lr_sig</span><span class="op">)</span> <span class="co"># interactions and their weights in the ligand-receptor + signaling network</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  from  to     weight
  &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 A1BG  ABCC6  0.422 
2 A1BG  ACE2   0.101 
3 A1BG  ADAM10 0.0970
4 A1BG  AGO1   0.0525
5 A1BG  AKT1   0.0855
6 A1BG  ANXA7  0.457 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">immune.combined.sct</span><span class="op">)</span> <span class="op">&lt;-</span><span class="va">immune.combined.sct</span><span class="op">$</span><span class="va">seurat_annotations</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">immune.combined.sct</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"integrated"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>在本文中，我们依赖于数据提供的筛选细胞注释。 我们将重点放在CD8 T细胞作为接收细胞，并寻找pDC、CD14 Mono、CD16 Mono、NK、B、DC、B Activated群集中的发送器配体。</p>
<p>我们定义接收细胞和发送器细胞群体。此外，我们还识别在接收细胞中表达的基因，这些基因也是我们知识库的一部分。</p>
<p>💡如果某个基因在数据库中没有被知道由任何配体调控，该方法将无法在其预测中使用它。</p>
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">receiver</span> <span class="op">=</span> <span class="st">"CD8 T"</span></span>
<span><span class="va">expressed_genes_receiver</span> <span class="op">=</span> <span class="fu">get_expressed_genes</span><span class="op">(</span><span class="va">receiver</span>, <span class="va">immune.combined.sct</span>, pct <span class="op">=</span> <span class="fl">0.10</span>, assay_oi<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="va">background_expressed_genes</span> <span class="op">=</span> <span class="va">expressed_genes_receiver</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">.</span><span class="op">[</span><span class="va">.</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ligand_target_matrix</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sender_celltypes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pDC"</span>, <span class="st">"CD14 Mono"</span> ,  <span class="st">"CD16 Mono"</span>,  <span class="st">"NK"</span>, <span class="st">"B"</span>, <span class="st">"DC"</span>, <span class="st">"B Activated"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">list_expressed_genes_sender</span> <span class="op">=</span> <span class="va">sender_celltypes</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">get_expressed_genes</span>, <span class="va">immune.combined.sct</span>, <span class="fl">0.10</span>, assay_oi<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> </span>
<span><span class="va">expressed_genes_sender</span> <span class="op">=</span> <span class="va">list_expressed_genes_sender</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>定义接收细胞中相对于样本处理的差异基因。</p>
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_obj_receiver</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, idents <span class="op">=</span> <span class="va">receiver</span><span class="op">)</span></span>
<span><span class="va">seurat_obj_receiver</span> <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">SetIdent</a></span><span class="op">(</span><span class="va">immune.combined.sct</span>, value <span class="op">=</span> <span class="va">seurat_obj_receiver</span><span class="op">[[</span><span class="st">"orig.ident"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">condition_oi</span> <span class="op">=</span> <span class="st">"IMMUNE_STIM"</span></span>
<span><span class="va">condition_reference</span> <span class="op">=</span> <span class="st">"IMMUNE_CTRL"</span> </span>
<span></span>
<span><span class="va">DE_table_receiver</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seurat_obj_receiver</span>, ident.1 <span class="op">=</span> <span class="va">condition_oi</span>, ident.2 <span class="op">=</span> <span class="va">condition_reference</span>, min.pct <span class="op">=</span> <span class="fl">0.10</span>, assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">geneset_oi</span> <span class="op">=</span> <span class="va">DE_table_receiver</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">p_val</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="va">geneset_oi</span> <span class="op">=</span> <span class="va">geneset_oi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">.</span><span class="op">[</span><span class="va">.</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ligand_target_matrix</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>定义在各自群集中表达的配体和受体基因。潜在的配体集被缩小为在接收细胞群集中表达的具有靶标的配体。</p>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ligands</span> <span class="op">=</span> <span class="va">lr_network</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">from</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">receptors</span> <span class="op">=</span> <span class="va">lr_network</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">to</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">expressed_ligands</span> <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="va">ligands</span>,<span class="va">expressed_genes_sender</span><span class="op">)</span></span>
<span><span class="va">expressed_receptors</span> <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="va">receptors</span>,<span class="va">expressed_genes_receiver</span><span class="op">)</span></span>
<span></span>
<span><span class="va">potential_ligands</span> <span class="op">=</span> <span class="va">lr_network</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">from</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">expressed_ligands</span> <span class="op">&amp;</span> <span class="va">to</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">expressed_receptors</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">from</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>NicheNet的配体活性分析根据接收细胞中存在的表达靶标基因的情况对配体进行排名。</p>
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ligand_activities = predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_ligands = potential_ligands)</span></span>
<span></span>
<span><span class="co"># ligand_activities = ligand_activities %&gt;% arrange(-pearson) %&gt;% mutate(rank = rank(desc(pearson)))</span></span>
<span><span class="co"># ligand_activities</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script type="text/javascript" src="//rf.revolvermaps.com/0/0/7.js?i=5j5b62dhqyd&amp;m=7&amp;c=ff0000&amp;cr1=ffffff&amp;sx=0" async="async"></script><script async="" src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><span id="busuanzi_container_site_pv">本站总访问量 <span id="busuanzi_value_site_pv"></span> 次（来源不蒜子按域名记录）</span>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/shixiangwang\.github\.io");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><script src="https://giscus.app/client.js" data-repo="shixiangwang/shixiangwang.github.io" data-repo-id="MDEwOlJlcG9zaXRvcnkzNDgxODQ1MTY=" data-category="General" data-category-id="DIC_kwDOFMDfxM4CVXEf" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><input type="hidden" id="giscus-base-theme" value="light"><input type="hidden" id="giscus-alt-theme" value="dark">
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Blog made with 💜 and <a href="https://quarto.org/">Quarto</a>, by Shixiang Wang. License: <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC BY-SA 2.0</a>.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/shixiangwang/">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/WangShxiang">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="mailto:shixiang1994wang@gmail.com">
      <i class="bi bi-envelope" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml">
      <i class="bi bi-rss" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>