<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.475">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Shixiang Wang">
<meta name="dcterms.date" content="2023-05-31">
<title>Reasoning - scRNA-seq_online 00：包准备</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><style>html{ scroll-behavior: smooth; }</style>
<!-- Primary Meta Tags --><title>Reasoning - ShixiangWang's blog</title>
<meta name="title" content="Reasoning - ShixiangWang's blog">
<meta name="description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<!-- Open Graph / Facebook --><meta property="og:type" content="website">
<meta property="og:url" content="https://shixiangwang.github.io/posts/welcome/">
<meta property="og:title" content="Reasoning - scRNA-seq_online 00：包准备">
<meta property="og:description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<meta property="og:image" content="https://avatars.githubusercontent.com/u/25057508?v=4">
<!-- Twitter --><meta property="twitter:card" content="summary_large_image">
<meta property="twitter:url" content="https://shixiangwang.github.io/posts/welcome/">
<meta property="twitter:title" content="Reasoning - ShixiangWang's blog">
<meta property="twitter:description" content="Reasoning is a blog dedicated to exploring the intersection of data science, bioinformatics, and personal insights.">
<meta property="twitter:image" content="https://avatars.githubusercontent.com/u/25057508?v=4">
<meta name="referrer" content="no-referrer">
<link rel="stylesheet" href="../../styles.css">
<meta name="twitter:title" content="Reasoning - scRNA-seq_online 00：包准备">
<meta name="twitter:description" content="安装 CRAN 包：">
<meta name="twitter:image" content="">
<meta name="twitter:creator" content="@WangShxiang">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Reasoning</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">
<li>
    <a class="dropdown-item" href="../../resource.html">
 <span class="dropdown-text">Code and Books</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://shixiangwang.github.io/cv-shixiang/">
 <span class="menu-text">Resume</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://shixiangwang.github.io/home/logo/qrcode.jpg"><i class="bi bi-wechat" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.zhihu.com/people/shixiangwang"><i class="bi bi-quora" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ShixiangWang"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/WangShxiang"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">scRNA-seq_online 00：包准备</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">bioinformatics</div>
                <div class="quarto-category">scRNA-seq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Shixiang Wang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">May 31, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#%E5%AE%89%E8%A3%85%E5%8C%85" id="toc-安装包" class="nav-link active" data-scroll-target="#%E5%AE%89%E8%A3%85%E5%8C%85">安装包</a></li>
  <li><a href="#%E6%B5%8B%E8%AF%95%E5%8C%85%E7%9A%84%E5%AE%89%E8%A3%85" id="toc-测试包的安装" class="nav-link" data-scroll-target="#%E6%B5%8B%E8%AF%95%E5%8C%85%E7%9A%84%E5%AE%89%E8%A3%85">测试包的安装</a></li>
  <li><a href="#%E6%9F%A5%E7%9C%8B%E4%BC%9A%E8%AF%9D%E4%BF%A1%E6%81%AF" id="toc-查看会话信息" class="nav-link" data-scroll-target="#%E6%9F%A5%E7%9C%8B%E4%BC%9A%E8%AF%9D%E4%BF%A1%E6%81%AF">查看会话信息</a></li>
  <li><a href="#%E9%98%85%E8%AF%BB" id="toc-阅读" class="nav-link" data-scroll-target="#%E9%98%85%E8%AF%BB">阅读</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><blockquote class="blockquote">
<p>参考资料：<a href="https://hbctraining.github.io/scRNA-seq_online/" class="uri">https://hbctraining.github.io/scRNA-seq_online/</a></p>
</blockquote>
<section id="安装包" class="level2"><h2 class="anchored" data-anchor-id="安装包">安装包</h2>
<p>安装 CRAN 包：</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tidyverse"</span>, <span class="st">"Matrix"</span>, <span class="st">"RCurl"</span>, <span class="st">"scales"</span>, <span class="st">"cowplot"</span>, <span class="st">"Seurat"</span>, <span class="st">"metap"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>安装 Bioc 包：</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># options(BioC_mirror="https://mirrors.tuna.tsinghua.edu.cn/bioconductor")</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span>, <span class="st">"ensembldb"</span>, <span class="st">"multtest"</span>, <span class="st">"glmGamPoi"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="测试包的安装" class="level2"><h2 class="anchored" data-anchor-id="测试包的安装">测试包的安装</h2>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">## The legacy packages maptools, rgdal, and rgeos, underpinning this package</span></span>
<span><span class="co">## will retire shortly. Please refer to R-spatial evolution reports on</span></span>
<span><span class="co">## https://r-spatial.org/r/2023/05/15/evolution4.html for details.</span></span>
<span><span class="co">## This package is now running under evolution status 0</span></span>
<span><span class="co">## Attaching SeuratObject</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="co">## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──</span></span>
<span><span class="co">## ✔ dplyr     1.1.2     ✔ readr     2.1.4</span></span>
<span><span class="co">## ✔ forcats   1.0.0     ✔ stringr   1.5.0</span></span>
<span><span class="co">## ✔ ggplot2   3.4.2     ✔ tibble    3.2.1</span></span>
<span><span class="co">## ✔ lubridate 1.9.2     ✔ tidyr     1.3.0</span></span>
<span><span class="co">## ✔ purrr     1.0.1</span></span>
<span><span class="co">## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span><span class="co">## ✖ dplyr::filter() masks stats::filter()</span></span>
<span><span class="co">## ✖ dplyr::lag()    masks stats::lag()</span></span>
<span><span class="co">## ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org">Matrix</a></span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Matrix'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:tidyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand, pack, unpack</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RCurl</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'RCurl'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:tidyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     complete</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org">scales</a></span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'scales'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:purrr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     discard</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:readr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     col_factor</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'cowplot'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:lubridate':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     stamp</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="co">## Loading required package: BiocGenerics</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:lubridate':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, union</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine, intersect, setdiff, union</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Loading required package: BiocFileCache</span></span>
<span><span class="co">## Loading required package: dbplyr</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dbplyr'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     ident, sql</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb">ensembldb</a></span><span class="op">)</span></span>
<span><span class="co">## Loading required package: GenomicRanges</span></span>
<span><span class="co">## Loading required package: stats4</span></span>
<span><span class="co">## Loading required package: S4Vectors</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:Matrix':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand, unname</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:lubridate':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     second, second&lt;-</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     first, rename</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:tidyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Loading required package: IRanges</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'IRanges'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:lubridate':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     %within%</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     collapse, desc, slice</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:purrr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     reduce</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Loading required package: GenomeInfoDb</span></span>
<span><span class="co">## Loading required package: GenomicFeatures</span></span>
<span><span class="co">## Loading required package: AnnotationDbi</span></span>
<span><span class="co">## Loading required package: Biobase</span></span>
<span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:AnnotationHub':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     cache</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'AnnotationDbi'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     select</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Loading required package: AnnotationFilter</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ensembldb'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="查看会话信息" class="level2"><h2 class="anchored" data-anchor-id="查看会话信息">查看会话信息</h2>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.2 (2022-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ensembldb_2.22.0        AnnotationFilter_1.22.0 GenomicFeatures_1.50.4 
 [4] AnnotationDbi_1.60.2    Biobase_2.58.0          GenomicRanges_1.50.2   
 [7] GenomeInfoDb_1.34.9     IRanges_2.32.0          S4Vectors_0.36.2       
[10] AnnotationHub_3.6.0     BiocFileCache_2.6.1     dbplyr_2.3.2           
[13] BiocGenerics_0.44.0     cowplot_1.1.1           scales_1.2.1           
[16] RCurl_1.98-1.12         Matrix_1.5-4.1          lubridate_1.9.2        
[19] forcats_1.0.0           stringr_1.5.0           dplyr_1.1.2            
[22] purrr_1.0.1             readr_2.1.4             tidyr_1.3.0            
[25] tibble_3.2.1            ggplot2_3.4.2           tidyverse_2.0.0        
[28] SeuratObject_4.1.3      Seurat_4.3.0           

loaded via a namespace (and not attached):
  [1] utf8_1.2.3                    spatstat.explore_3.2-1       
  [3] reticulate_1.28               tidyselect_1.2.0             
  [5] RSQLite_2.3.1                 htmlwidgets_1.6.2            
  [7] grid_4.2.2                    BiocParallel_1.32.6          
  [9] Rtsne_0.16                    munsell_0.5.0                
 [11] codetools_0.2-19              ica_1.0-3                    
 [13] future_1.32.0                 miniUI_0.1.1.1               
 [15] withr_2.5.0                   spatstat.random_3.1-5        
 [17] colorspace_2.1-0              progressr_0.13.0             
 [19] filelock_1.0.2                knitr_1.43                   
 [21] rstudioapi_0.14               ROCR_1.0-11                  
 [23] tensor_1.5                    listenv_0.9.0                
 [25] MatrixGenerics_1.10.0         GenomeInfoDbData_1.2.9       
 [27] polyclip_1.10-4               bit64_4.0.5                  
 [29] parallelly_1.36.0             vctrs_0.6.2                  
 [31] generics_0.1.3                xfun_0.39                    
 [33] timechange_0.2.0              R6_2.5.1                     
 [35] bitops_1.0-7                  spatstat.utils_3.0-3         
 [37] cachem_1.0.8                  DelayedArray_0.24.0          
 [39] promises_1.2.0.1              BiocIO_1.8.0                 
 [41] gtable_0.3.3                  globals_0.16.2               
 [43] goftest_1.2-3                 rlang_1.1.1                  
 [45] splines_4.2.2                 rtracklayer_1.58.0           
 [47] lazyeval_0.2.2                spatstat.geom_3.2-1          
 [49] BiocManager_1.30.20           yaml_2.3.7                   
 [51] reshape2_1.4.4                abind_1.4-5                  
 [53] httpuv_1.6.11                 tools_4.2.2                  
 [55] ellipsis_0.3.2                RColorBrewer_1.1-3           
 [57] ggridges_0.5.4                Rcpp_1.0.10                  
 [59] plyr_1.8.8                    progress_1.2.2               
 [61] zlibbioc_1.44.0               prettyunits_1.1.1            
 [63] deldir_1.0-9                  pbapply_1.7-0                
 [65] zoo_1.8-12                    SummarizedExperiment_1.28.0  
 [67] ggrepel_0.9.3                 cluster_2.1.4                
 [69] magrittr_2.0.3                data.table_1.14.8            
 [71] scattermore_1.1               lmtest_0.9-40                
 [73] RANN_2.6.1                    ProtGenerics_1.30.0          
 [75] fitdistrplus_1.1-11           matrixStats_0.63.0           
 [77] hms_1.1.3                     patchwork_1.1.2              
 [79] mime_0.12                     evaluate_0.21                
 [81] xtable_1.8-4                  XML_3.99-0.14                
 [83] gridExtra_2.3                 compiler_4.2.2               
 [85] biomaRt_2.54.1                KernSmooth_2.23-21           
 [87] crayon_1.5.2                  htmltools_0.5.5              
 [89] later_1.3.1                   tzdb_0.4.0                   
 [91] DBI_1.1.3                     MASS_7.3-60                  
 [93] rappdirs_0.3.3                cli_3.6.1                    
 [95] parallel_4.2.2                igraph_1.4.3                 
 [97] pkgconfig_2.0.3               GenomicAlignments_1.34.1     
 [99] sp_1.6-1                      plotly_4.10.1                
[101] spatstat.sparse_3.0-1         xml2_1.3.4                   
[103] XVector_0.38.0                digest_0.6.31                
[105] sctransform_0.3.5             RcppAnnoy_0.0.20             
[107] spatstat.data_3.0-1           Biostrings_2.66.0            
[109] rmarkdown_2.21                leiden_0.4.3                 
[111] uwot_0.1.14                   restfulr_0.0.15              
[113] curl_5.0.0                    shiny_1.7.4                  
[115] Rsamtools_2.14.0              rjson_0.2.21                 
[117] lifecycle_1.0.3               nlme_3.1-162                 
[119] jsonlite_1.8.4                viridisLite_0.4.2            
[121] fansi_1.0.4                   pillar_1.9.0                 
[123] lattice_0.21-8                KEGGREST_1.38.0              
[125] fastmap_1.1.1                 httr_1.4.6                   
[127] survival_3.5-5                interactiveDisplayBase_1.36.0
[129] glue_1.6.2                    png_0.1-8                    
[131] BiocVersion_3.16.0            bit_4.0.5                    
[133] stringi_1.7.12                blob_1.2.4                   
[135] memoise_2.0.1                 irlba_2.3.5.1                
[137] future.apply_1.11.0          </code></pre>
</div>
</div>
</section><section id="阅读" class="level2"><h2 class="anchored" data-anchor-id="阅读">阅读</h2>
<p>接下来就可以阅读以下两份非常好的材料了解单细胞的预备知识。</p>
<ul>
<li><a href="https://hbctraining.github.io/scRNA-seq_online/lessons/01_intro_to_scRNA-seq.html">Introduction to single-cell RNA-seq</a></li>
<li><a href="https://hbctraining.github.io/scRNA-seq_online/lessons/02_SC_generation_of_count_matrix.html">Generation of count matrix</a></li>
</ul>


</section></main><!-- /main --><script type="text/javascript" src="//rf.revolvermaps.com/0/0/7.js?i=5j5b62dhqyd&amp;m=7&amp;c=ff0000&amp;cr1=ffffff&amp;sx=0" async="async"></script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><script src="https://giscus.app/client.js" data-repo="shixiangwang/shixiangwang.github.io" data-repo-id="MDEwOlJlcG9zaXRvcnkzNDgxODQ1MTY=" data-category="General" data-category-id="DIC_kwDOFMDfxM4CVXEf" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Blog made with 💜 and <a href="https://quarto.org/">Quarto</a>, by Shixiang Wang. License: <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC BY-SA 2.0</a>.</div>   
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/shixiangwang/">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/WangShxiang">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="mailto:shixiang1994wang@gmail.com">
      <i class="bi bi-envelope" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml">
      <i class="bi bi-rss" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>